### PR TITLE
Kraken Private Futures API Client

### PR DESCRIPTION
**Pull request by rewong03**
### Description
wrapper client and containers for the following endpoints for kraken's future rest api:
- https://docs.kraken.com/api/docs/futures-api/trading/get-open-positions
- https://docs.kraken.com/api/docs/futures-api/trading/send-batch-order
- https://docs.kraken.com/api/docs/futures-api/trading/cancel-all-orders
- https://docs.kraken.com/api/docs/futures-api/trading/cancel-order
- https://docs.kraken.com/api/docs/futures-api/trading/edit-order-spring
- https://docs.kraken.com/api/docs/futures-api/trading/get-open-orders
- https://docs.kraken.com/api/docs/futures-api/trading/send-order
- https://docs.kraken.com/api/docs/futures-api/trading/get-order-status

### Testing
Integration tests with the kraken test environment using a test account (keys added to the repo already) + unit testing for helper functions and serialization

### Impact
probably nothing rn

### Other
all test runners use the same api credentials, so if there are a ton running at once its possible to get rate limited (but re-running tests should fix the issue) https://docs.futures.kraken.com/#http-api-http-api-introduction-request-limits

### CODE REVIEW

File: src/pysrc/adapters/kraken/future/containers.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            order: Order,
+            process_before: str,
+            trailing_stop_max_deviation: Optional[bool],
+            trailing_stop_deviation_unit: Optional[PriceUnit],
+            limit_price_offset_value: Optional[float],
+            limit_price_offset_unit: Optional[PriceUnit]
+    ):
+        pass
+
+class PositionSide(Enum):
```
Comment (echavemann) [inline src/pysrc/adapters/kraken/future/containers.py:64]: we have an enum in util.types you can probably use here

Comment (echavemann) [reply src/pysrc/adapters/kraken/future/containers.py:64]: if you think it needs to be different for held positions, I would add that to types

File: src/pysrc/adapters/kraken/future/containers.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    SPOT = 1
+    LAST = 2
+
+
+class Order:
+    def __init__(
+        self,
+        symbol: str,
+        side: OrderSide,
+        size: Optional[float] = None,
```
Comment (echavemann) [inline src/pysrc/adapters/kraken/future/containers.py:42]: how is this optional

Comment (rewong03) [reply src/pysrc/adapters/kraken/future/containers.py:42]: "quantity" listed as numbernull under https://docs.kraken.com/api/docs/futures-api/trading/get-order-status/

File: src/pysrc/adapters/kraken/future/utils.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        case _:
+            raise ValueError(f"Can't convert '{s}' to OrderSide")
+
+
+def order_side_to_str(o: OrderSide) -> str:
+    match o:
+        case OrderSide.BID:
+            return "buy"
+        case OrderSide.ASK:
+            return "sell"
```
Comment (echavemann) [inline src/pysrc/adapters/kraken/future/utils.py:60]: case _

File: src/pysrc/adapters/kraken/future/utils.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+def str_to_price_unit(s: str) -> PriceUnit:
+    return PriceUnit[s.upper()]
+
+
+def price_unit_to_str(p: PriceUnit) -> str:
+    return p.name.lower()
+
+
+def remove_empty_values(d: dict[str, Any]) -> dict[str, Any]:
```
Comment (echavemann) [inline src/pysrc/adapters/kraken/future/utils.py:71]: bad var name, also can these be str, object?

Comment (echavemann) [reply src/pysrc/adapters/kraken/future/utils.py:71]: err fn name, but I dont think u can name them dict either

File: src/pysrc/test/integration/adapters/kraken/future/test_kraken_future_client.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+    client = KrakenFutureClient(
+        public_key=public_key, private_key=private_key, use_live_api=False
+    )
+
+    return client
+
+
+def test_place_and_cancel(client: KrakenFutureClient) -> None:
+    # cancel any existing orders
```
Comment (echavemann) [inline src/pysrc/test/integration/adapters/kraken/future/test_kraken_future_client.py:35]: del

Comment (echavemann) [reply src/pysrc/test/integration/adapters/kraken/future/test_kraken_future_client.py:35]: remove all these comments lol

File: src/pysrc/adapters/kraken/future/containers.py (hunk 1 above)
Comment (rewong03) [reply src/pysrc/adapters/kraken/future/containers.py]: moved in 0df69dfe936d70fbac0b33a6aa8e443e3f90fa08

File: src/pysrc/adapters/kraken/future/containers.py (hunk 2 above)
Comment (rewong03) [reply src/pysrc/adapters/kraken/future/containers.py:42]: good to resolve?

File: src/pysrc/adapters/kraken/future/utils.py (hunk 4 above)
Comment (rewong03) [reply src/pysrc/adapters/kraken/future/utils.py]: renamed to `remove_empty_values_from_dict` in 0df69dfe936d70fbac0b33a6aa8e443e3f90fa08

File: src/pysrc/test/integration/adapters/kraken/future/test_kraken_future_client.py (hunk 5 above)
Comment (rewong03) [reply src/pysrc/test/integration/adapters/kraken/future/test_kraken_future_client.py]: removed in 0df69dfe936d70fbac0b33a6aa8e443e3f90fa08

**Comment by rewong03** [comment]
any reason why `safe_at`/`get_single` are in `adapters/kraken/future/containers.py` and not anywhere higher?

**Comment by rewong03** [comment]
also let's try to merge #40 first before this

**Comment by echavemann** [comment]
> any reason why `safe_at`/`get_single` are in `adapters/kraken/future/containers.py` and not anywhere higher? @echavemann @minnce

mb didnt realize, this needs to get patched

**Pull request merged**